optaplanner-core/src/main/java/org/optaplanner/core/api/score/stream/uni/UniConstraintStream.java

```diff
@@ -180,7 +190,332 @@
      */
     default <B> BiConstraintStream<A, B> join(Class<B> otherClass, BiJoiner<A, B>... joiners) {
-        return join(otherClass, AbstractBiJoiner.merge(joiners));
+        int joinerCount = joiners.length;
+        int indexOfFirstFilter = -1;
+        for (int i = 0; i < joinerCount; i++) {
+            BiJoiner<A, B> joiner = joiners[i];
+            boolean hasAFilter = indexOfFirstFilter >= 0;
+            if (joiner instanceof FilteringBiJoiner) {
+                if (!hasAFilter) { // From now on, we only allow filtering joiners.
+                    indexOfFirstFilter = i;
+                }
+            } else if (hasAFilter) {
+                throw new IllegalStateException("Indexing joiner (" + joiner + ") must not follow a filtering joiner ("
+                        + joiners[indexOfFirstFilter] + ").");
+            }
+        }
+        if (indexOfFirstFilter < 0) { // Only found indexing joiners.
+            return join(otherClass, AbstractBiJoiner.merge(joiners));
+        }
+        // Assemble the join stream.
+        BiConstraintStream<A, B> joined = indexOfFirstFilter == 0 ?
+                join(otherClass) :
+                join(otherClass, Arrays.copyOf(joiners, indexOfFirstFilter));
+        // We merge all filters into one, so that we don't pay the penalty for lack of indexing more than once.
+        BiPredicate<A, B> resultingFilter = (a, b) -> true;
+        for (int i = indexOfFirstFilter; i < joinerCount; i++) {
+            FilteringBiJoiner<A, B> filteringJoiner = (FilteringBiJoiner<A, B>) joiners[i];
+            resultingFilter = resultingFilter.and(filteringJoiner.getFilter());
+        }
+        return joined.filter(resultingFilter);
+    }
+
+    /**
+     * Create a new {@link BiConstraintStream} for every two different A.
+     * <p>
+     * Important: This is faster and more scalable than a {@link #joinOther(UniConstraintStream) joinOther}
+     * followed by a {@link BiConstraintStream#filter(BiPredicate) filter},
+     * because it applies hashing and/or indexing on the properties,
+     * so it doesn't create nor checks every combination of A and B.
+     * @param otherStream never null
+     * @return a stream that matches every combination of A and a different A
+     */
+    default BiConstraintStream<A, A> joinOther(UniConstraintStream<A> otherStream) {
+        return join(otherStream, Joiners.notEqual());
+    }
+
+    /**
+     * Create a new {@link BiConstraintStream} for every combination of A and a different A for which the
+     * {@link BiJoiner} is true (for the properties it extracts from both facts).
+     * <p>
+     * Important: This is faster and more scalable than a {@link #joinOther(UniConstraintStream) joinOther}
+     * followed by a {@link BiConstraintStream#filter(BiPredicate) filter},
+     * because it applies hashing and/or indexing on the properties,
+     * so it doesn't create nor checks every combination of A and B.
```

Comment:
join(Class) and join(Class, joiner) are very commonly used, so they don't use "As defined by" in the javadocs. They are an exception, because using "as defined by" reduces maintenance.

I'd argue that "joinOther(Class)" can use "As defined by join(Class)" and then something like "With the additional joiner than they are not equal" (but better written).
Same for "joinOther(Class, Joiner)" => "As defined by join(Class, Joiner) with the additional joiner than they are not equal"
But indeed for joinOther(Class, Joiner, Joiner) and more, I 'd keep using "As defined by joinOther(Class, Joiner, Joiner)" (so no need of "with the additional ..." because it references joinOther, not join).

Wdyt?

Reply:
We've just agreed during our call that we're removing `joinOther` altogether. :-)